### ISSUE TITLE
JUGGLING ASYNC - Trying to wrap my mind around asynchronous programming.

### ISSUE THREAD

**Issue by theCard** [nodeschool/discussions#69]
I slightly changed the solution by replacing the httpGet function with a for loop. I was just wondering, why doesn't this work? Does calling httpGet somehow synchronize the method calls?

```
var http = require('http')
var bl = require('bl')

var count = 0;
var arrhold = []

for (var i = 2; i < 5; i++) 
http.get(process.argv[i], function (response) {
  response.pipe(bl(function (err, data) {
    if (err)
     return console.error(data)      

     arrhold[count] = data.toString()
      count++

     if (count == 3)
     printResults()

  }))  
})


function printResults() {
for (var i = 0; i < 3; i++)
console.log(arrhold[i])
}
```

**Comment by JohnnyNiu** [comment]
Thank you for your clear explanation. @magwitch

**Comment by chanhannigan** [comment]
All these comments helped immensely. I didn't think there would be such a variety of working solutions. Thanks guys, and here's where my mind took me.
```javascript
var http = require('http');

var urlObject = {
	obj1: new getData(process.argv[2]),
	obj2: new getData(process.argv[3]),
	obj3: new getData(process.argv[4])
};

var finishedCount = 0;

function getData(url) {
	
	this.contents = '';

    var that      = this;
	var options   = {
		"data" : function(data) { that.contents += data; },
		"end"  : function()     { finish(); }
	};

	var finish = function() {
		finishedCount++;
		if( finishedCount === 3){
			for(i in urlObject) {
				console.log(urlObject[i].contents);
			}
		}
	};

	http.get(url, function(response) {
		for(key in options) {
			response.on(key, options[key]);
		};
	});
};
```

**Comment by martinheidegger** [comment]
@chanhannigan 

```JavaScript
var urlObject = {
    obj1: new getData(process.argv[2]),
    obj2: new getData(process.argv[3]),
    obj3: new getData(process.argv[4])
};
```

`new` in your case will create simple objects that don't do anything.

```JavaScript
new getData(process.argv[2]),
    obj2: new getData(process.argv[3]),
    obj3: new getData(process.argv[4])
};
```

This would work too.

**Comment by chanhannigan** [comment]
Ahha! Thanks for the pro-tip @martinheidegger

**Comment by martinheidegger** [comment]
Oh dear, my bad. I should have rechecked what I wrote before I submitted it (was interrupted inbetween). You use a coding style that is completely unfamiliar to me (and is hard to read in general imho) and I mistook some things. Checking more carefully I see where I went wrong and it seems you have a perfectly fine implementation.

**Comment by vtoomas** [comment]
Hi, 

Could someone help me to understand what is wrong with my following solution?

```javascript
var http = require('http');
var urls = [process.argv[2], process.argv[3], process.argv[4]];
var count = 0; 
var printArr = ["","",""];

function runUrl(i) {
	var url = urls[i];
	var req = http.get(url);
	req.on('response', function(resp) {
		resp.on('data', function(data) {
			printArr[i] += data.toString(); 
			
		});
		resp.on('end', function() {
			count++;
			if (count == urls.length - 1) {
				PrintInfo();  
			}
		})
	});
	req.on('error', function(e) {
	  console.log('problem with request: ' + e.message);
	});
	req.end();
}

function PrintInfo() {
	printArr.forEach(function(urldata) {
		console.log(urldata);
	})
}
for (i=0;i<urls.length;i++) {
	runUrl(i);
}
```

I always get one response incomplete like following: 

1.  ACTUAL:    "As dry as a cream heaps she'll be right squizz. She'll be right
bradman flamin flat out like a pint. "
1.  EXPECTED:  "As dry as a cream heaps she'll be right squizz. She'll be right
bradman flamin flat out like a pint. "

2.  ACTUAL:    "Come a aussie salute to we're going two up. He hasn't got a tuck
er-bag where as busy as a sickie. Come a struth when we're going ute. "
2.  EXPECTED:  "Come a aussie salute to we're going two up. He hasn't got a tuck
er-bag where as busy as a sickie. Come a struth when we're going ute. "

3.  ACTUAL:    "Lets get some kindie when he's got a massive trackies. Mad as a
captain cook flamin as cunning as a chuck a yewy. Stands out like a "
3.  EXPECTED:  "Lets get some kindie when he's got a massive trackies. Mad as a
captain cook flamin as cunning as a chuck a yewy. Stands out like a dob with bui
lt like a doovalacky. Grab us a down under my grab us a g'day. She'll be right c
ut lunch commando with lets get some true blue. "

**Comment by Sheffyp** [comment]
Hi guys,

This code passed...

'''
var http = require('http');
var url1 = process.argv[2];
var url2 = process.argv[3];
var url3 = process.argv[4];
pagedata1="";
pagedata2="";
pagedata3="";
callbackcounter=0;

var printout = function (){
  console.log(pagedata1);
  console.log(pagedata2);
  console.log(pagedata3);
};


http.get(url1, function(response){
  response.setEncoding("utf8");
  response.on("data", function(data) {
    pagedata1+=data;
});
  response.on("end", function(){
    callbackcounter+=1;
    if (callbackcounter==3){
      printout();
}});
});

http.get(url2, function(response){
  response.setEncoding("utf8");
  response.on("data", function(data) {
    pagedata2+=data;
});
  response.on("end", function(){
    callbackcounter+=1;
    if (callbackcounter==3){
      printout();
}});
});

http.get(url3, function(response){
  response.setEncoding("utf8");
  response.on("data", function(data) {
    pagedata3+=data;
});
  response.on("end", function(){
    callbackcounter+=1;
    if (callbackcounter==3){
      printout();
};
});
});
'''

Yes I know its pretty awful - no loops and definitely not following the "dont repeat yourself" mantra, but the reason I post it here, is that its simple and acts as an example of monitoring the callbacks, in its most simple form.

**Comment by jmdecombe** [comment]
```javascript
const http = require('http')
const bl = require('bl')
const source = process.argv.slice(2)
var count = 0
var content = []
function printContent() {
	for (var i = 0; i < content.length; i++) {
		console.log(source[i] + ' = ' + content[i])
	}
}
function getContent(source, index) {
	http.get(source[index], function(response) {
		response.pipe(bl(function (error, data) {
			if (error) {
				return console.error(error)
			}
			content[index] = data.toString();
			if (++count == source.length) {
				printContent();
			}
		}))
	})
}
for (var i = 0; i < source.length; i++) {
	getContent(source, i)
}
```

**Comment by MatejMatula** [comment]
Hi , i found nodeschool pretty broken...
this code


var http=require("http");
var urls=process.argv.slice(2,process.argv.length);
var count=0;
var junge=[];
urls.map(function(url,index){
http.get(url,function(response){
    var str="";
    response.setEncoding("utf-8");

    response.on("data",function(data){str=str+data})
    response.on("end",function(){
        junge.push(str);
        count++;
        if(index==urls.length-1) junge.map(function(v){console.log(v)})

    });

});})

worked , now it does not work...

**Comment by cameronlockey** [comment]
I borrowed a little from some of your solutions, but then made it into an object-based module (just for practice, mostly). The `getURLs` method accepts an array of URLs and figures out how many callbacks it needs to wait for before it prints. 

    var http = require('http')
    var concat = require('concat-stream')

    var httpStreamResponses = {

      /**
       * Holder array for return data
       * @type array
       */
      pages: [],

      /**
       * URLs to be requested
       * @type array
       */
      urls: [],

      /**
       * # of requests waiting to finish
       * @type array
       */
      waiting: 0,

      /*
       * Get the HTTP response for all urls provided
       *
       * @param array urls
       * @param function callback
       * @return mixed Callback or default (void) this.complete()
       */
      getURLs: function(urls, callback) {

        var _this = this

        if (urls.length) {
          // set urls and waiting count
          this.urls = options.urls
          this.waiting = this.urls.length

          // start getting responses
          this.urls.map(function(url,idx){
            _this.httpGet(url, function(data){
              data = data.toString()
              _this.pages[idx] = data
              _this.waiting--
              if (!_this.waiting) {
                if (callback)
                  return callback(_this.pages)
                else
                  _this.complete()
              }
            })
          })
        } else {
          console.error("You must provide an array of URLs to get.")
        }

      },

      /**
       * Get the concatenated HTTP string (utf8) response and feed it into a Callback
       * @param  string   url
       * @param  function callback
       * @return void
       */
      httpGet: function (url, callback) {
        var _this = this
        http.get(url, function(rs){
          rs.setEncoding('utf8')
          rs.pipe(concat(function(data){
            callback(data)
          }))
        })
      },

      /**
       * Default callback method for getURLs().
       * Logs the response of each URL to the console.
       * @return void
       */
      complete: function() {
        this.pages.map(function(page){
           console.log(page)
        })
      }
    }

    module.exports = httpStreamResponses

And the program invoking it would be:

    var httpStream = require('./http-stream-responses.js')

    var args = process.argv.slice(2, process.argv.length)
    httpStream.getURLs({urls:args})

And if you wanted you could write your own callback into `getURLs()`:

    httpStream.getURLs({urls:args}, function(){ 
      ... do whatever besides log to the console... 
    })

**Comment by rhea-capiot** [comment]
Hi can anybody tell me why this code doesnt work??

var net = require('net')

function zero(i) {
  return (i < 10 ? '0' : '') + i
}
function now () {
  var d = new Date()
  return d.getFullYear() + '-' + zero(d.getMonth()) + '-'
    + zero(d.getDay()) + ' ' + zero(d.getHours()) + ':'
    + zero(d.getMinutes())
}

var server = net.createServer(function (socket) {
  socket.end(now() + '\n')
}).listen(8000)

**Comment by rhea-capiot** [comment]
var http = require('http')
var bl = require('bl')
var data=[]
var i=2;
var k=0
var count=0;
var pop;
for(i; i<=4; i++)
{
http.get(process.argv[i], function (request) {
  request.pipe(bl(function (err, dataa) {
    if (err)
      return console.error(dataa);
	//console.log(dataa);
    data[i-2] =dataa.toString();
	count ++
    //console.log( data[i-2])
	if(count==3)
	{	myfunc()
	}
  }))  
})
}

 function myfunc(){
	for(k=0; k<3; k++){
		console.log(data[k])}
}

**Comment by BrandTurner** [comment]
@katjad I had the same problem with the results. I think I went a little overboard by creating a custom event emitter instead of just using a counter variable, but it was instructive. Here's the code:

```javascript
var http 		= require('http');
var util 		= require('util');
var urls		= process.argv.slice(2);
var bl 			= require('bl');
var results 	= [];
var ticks 		= 0;
EventEmitter 	= require('events').EventEmitter

function printResults() {
	for (var i=0;i<results.length;i++) {
		console.log(results[i]);
	}
}
var fetch = function() {
	var self = this;
	urls.map(function(url, index) {
		http.get(url, function(response) {
			response.setEncoding('utf8');
			response.on('error', function(error){ console.error(error) });
			response.pipe(bl(function (error, data) {
				if (error) return console.error(error);
				results[index] = data.toString();				
				self.emit('tick');
			}));
		})
	})
}

util.inherits(fetch, EventEmitter);
var foo = new fetch();

foo.on("tick", function() {
	if (++ticks === 3) {
		printResults();
	}
});
```

**Comment by mike-sudyn** [comment]
Yet another solution:
```javascript
var http = require('http'),
    urls = [ process.argv[2], process.argv[3], process.argv[4]];

function request ( urls, callback ) {
    var waiting = urls.length,
        result = {};
    urls.map( function( url ){
        http.get( url, function ( response ){
            var collectedData = '';

            response.setEncoding('utf8');
            response.on('data', function( data ){ collectedData += data; });
            response.on('end', function(){
                waiting--;
                result[url] = collectedData;
                if( !waiting ){ callback( result ); }
            });
        });
    });
}

request( urls, function( data ){
    urls.map( function( url ) {
        console.log( data[url] );
    });
});

**Comment by booshay** [comment]
var http=require('http');
var concat=require('concat-stream');
var arry=[];
var counter=0;
 
function buildOutput(index){  
 http.get(process.argv[index+2], function(response){
  response.setEncoding('utf8');
  var output=concat(function(response){
   arry[index]=(response);
   counter++;
   if (counter==3){
    for (var i=0; i<arry.length; i++){
    console.log(arry[i]);
    }
   }
  });
  response.pipe(output);
 });
}

for(var i=0; i<3; i++){
    buildOutput(i);
}

**Comment by jjlucas76** [comment]
My solution:
var http = require('http'),
x = 0,
uri = [];
uri[0] = process.argv[2];
uri[1] = process.argv[3];
uri[2] = process.argv[4];

var array = {array0: [], array1: [], array2: []};

http.get(uri[x], whendata);

function whendata(response) {
    response.setEncoding('utf8');
    response.on('data', addToArray);
    response.on('error', callerror);
    response.on('end', count_response);
}
function display() {
    
    for (var ar in array){
        var stringa = [];
        var decoded = array[ar].toString();
        for (var t = 0; t <= decoded.length - 1; t++) {
            if (decoded[t] !== ',') { stringa += decoded[t]}
        }
        console.log(stringa);
    }
}


function callerror(error) {
    console.error(error);
}

function addToArray(data) {
    array['array'+x].push(data);
}

function count_response() {
    x = x+1;
    if (x===3) { 
        display();return; 
    }
    else {
        http.get(uri[x], whendata);
    }
}

**Comment by benassi84** [comment]
Submission results match expected ;)
```javascript
var http = require('http');
var bl = require('bl');
var urls = process.argv.slice(2, process.argv.length);
var urlsCount = urls.length;
var arrayStream = [];
var countCallback = 0;

printData = function () {
    arrayStream.map(function (data, index) {
        console.log(data);
    });
};
parseUrls = function (url, index) {
    http.get(url, function (response) {
        response.setEncoding('utf8');
        response.pipe(bl(function (err, data) {
            if (err)
                console.log(err);
            arrayStream[index] = data.toString();
            countCallback++; 
            if (countCallback === urlsCount)
                printData();
        }));
    });
};
urls.map(function (data, index) {
    parseUrls(data,index);
});
```

**Comment by pseudecoder** [comment]
Solution w/o ext module

```javascript
const http = require('http');

var count = 0;
var results = [];
var argv = process.argv;


function printResult() {
  for (var i =0; i < argv.length - 2; i++) {
    console.log(results[i]);
  }
}

function httpGet(index) {
  http.get(argv[index + 2], function(res) {
    res.setEncoding('utf8');
    
    results[index] = '';
    
    res.on('data', function(data) {
      results[index] += data;
    });

    res.on('end', function(){
      count++;

      if (count == 3) {
        printResult();
      }
    });

    res.on('error', function(err){
      return console.error(err);
    });
  });
}

for (var i = 0; i < argv.length - 2; i++) {
  httpGet(i);
}
```

**Comment by skoocda** [comment]
```
var http = require('http')
var count = 0
var results = []

function httpGet(index){
  http.get(process.argv[2+index], function(response){
    var stringOut = []
    response.setEncoding('utf8')
    response.on('error', console.error)
    response.on('data', function(data){
      stringOut.push(data)
    })
    response.on('end', function(){
      results[index] = stringOut.join('')
      count++

      if (count == 3)
        printResults()
    })
  })
}

function printResults(){
  for(var i=0;i<3;i++)
  console.log(results[i])
}

for (var k = 0; k < 3; k++)
  httpGet(k)

```

**Comment by taufiqahmed** [comment]
This worked beautifully:

```
var http = require("http");

var urlList = [process.argv[2], process.argv[3], process.argv[4]];

urlList.forEach(function(url, index){
  http.get(url, function(response){
    var responseBody = '';
    response.setEncoding('utf8');
    response.on('data', function(data){
      responseBody += data;
    });
    response.on('end', function() {
      console.log(responseBody);
    });
  });
});
```

**Comment by maxww** [comment]
I don't understand why this won't work...

```
var http = require("http");
var bl = require("bl");
var urls = [process.argv[2], process.argv[3], process.argv[4]];
var result = [];

urls.forEach(function (url, index) {
	http.get(url, function (res) {
		res.pipe(bl(function (err, data) {
			result[index] = data.toString();
		}))
	})
})

for (var i = 0; i < result.length; i++) {
	console.log(result[i]);
}
```

**Comment by noffle** [comment]
@maxww: you're close! However, when you run asynchronous functions (like `http.get()`) you aren't given any guarantees about when it will return. So, starting 3 such functions leaves you with no certainty about what order the replies will come back in. You can imagine that if you were pinging `google.com` it might respond faster than say `eight45.net`. Think about how you might ensure the results are collected (or arranged upon completion) to the ordering you care about.

**Comment by mmutuyu** [comment]
my solution using recursion:
```
var http = require('http');

var q = [process.argv[2], process.argv[3], process.argv[4]];

(function rec(url) {
  http.get(url, function (response) {
    var res = '';
    response.setEncoding('utf8');
    response.on ("data", function (data) {
      res += data;
    });
    response.on ("end", function() {
      console.log(res);
      if(url) rec(q.shift());
    });
  });
})(q.shift());
```

**Comment by crazyyi** [comment]
Here is my solution without using third party modules:

```
var http = require('http');
var urls = process.argv.splice(2);
var count = 0;

var results = [];

function getData(index) {
		http.get(urls[index], (res) => {
			var data ='';
			res.setEncoding('utf8');
			res.on('data', function(chunk) {
				data += chunk;
			});

			res.on('end', function() {
				results[index] = data;
				count++;
				if (count === urls.length) {
					for(var i = 0; i < urls.length; i++) {
						console.log(results[i]);
					}
				};
				
			});
			
		});
}

for (var i = 0; i < urls.length; i++) {
	getData(i);
}
```

**Comment by AndreAmo** [comment]
Hi there, I've solved the exercise thanks to the previous messages but I still have a doubt and if anyone could give me any insight would be much appreciated, assuming that this is the solution I used:

```
var http = require('http');
var concatStream = require('concat-stream');
var count = 0;

var answers = [];

function printRes () {
	answers.forEach(function (el) {
		console.log(el);
	});
}
```
Why this one works:
```
function getData (index) {
    http.get(process.argv[index + 2], function (res) {
        res.on('error', console.error);
            res.pipe(concatStream(function (data) {
                answers[count] = data.toString();
                count++;
                if (count === 3) printRes();
            }));
        }).on('error', console.error);
}

for (var i = 0; i < 3; i++) {
    getData(i);
}
```
and this one arranges the results in the wrong order:

```
for (var i = 0; i < 3; i++) {
	http.get(process.argv[index + 2], function (res) {
		res.on('error', console.error);
		res.pipe(concatStream(function (data) {
			answers[count] = data.toString();
			count++;
			if (count === 3) printRes();
		}));
	}).on('error', console.error);
}
```
thanks in advance for the reply.

**Comment by ishtiaque23** [comment]
Before going to the "Answer" part, let us go through the 2 code samples you shared.

First I will add some console.log line to your code(in both samples) right after the **count++** statement. 
_Note: These are for debugging purpose, not part of the solution at all._

```javascript
console.log('current index: ', index)
console.log('our counter(here count) for results array: ', count)
console.log('actual numbers of item in results array: ', results.length)
console.log('------------------------------------')
```

I want to do simulation with both of them.

First the correct code (result taken from my console):
```
$ learnyounode run webclientsync.js
current index:  1
our counter(here count) for results array:  1
actual numbers of item in results array:  2
------------------------------------
current index:  2
our counter(here count) for results array:  2
actual numbers of item in results array:  3
------------------------------------
current index:  0
our counter(here count) for results array:  3
actual numbers of item in results array:  3
------------------------------------
As stands out like avos piece of piss mad as a cobber. Built like a dag where stands out like a divvy van. As cunning as a scratchy bloody he hasn't got a aussie salute. He's got a massive cook piece of piss as dry as a knock. 
Grab us a porky and flat out like a chuck a sickie. She'll be right buck's night no dramas watch out for the fair go. Come a counter meal how shazza got us some buckley's chance. He hasn't got a quid heaps he's got a massive figjam. 
She'll be right doovalacky bloody lets get some jumbuck. He hasn't got a dropkick and he's got a massive bingle. He's got a massive dunny rat mate as cross as a vee dub. Get a dog up ya metho to mad as a plonk. Trent from punchy vb where mad as a slaps.
```

Everything is right about this simulation: 
* the **index** is passed properly, though in an unsorted order (1, 2 and 0). But thats expected as we know the **http.get** response will be async and callback will kick in for whichever appears first. 
* The **count** is also updating properly.
* The **results** array **length** is a tricky one but that is also correct. After all it finally returns 3 items.
* Also you can see all the values of results array are printing out properly in the end.

Now let's do the simulation with the wrong code:
```
$ learnyounode run webclientsync.js
current index:  3
our counter(here count) for results array:  1
actual numbers of item in results array:  4
------------------------------------
current index:  3
our counter(here count) for results array:  2
actual numbers of item in results array:  4
------------------------------------
current index:  3
our counter(here count) for results array:  3
actual numbers of item in results array:  4
------------------------------------
undefined
undefined
undefined
```
Well, well, well, whats going on here?
* In all three iteration the **index** is a constant 3. Thats not right
* The **count** is behaving properly though
* The **results** array **length** is another faulty one and its a stubborn **4**
* Final printout of the results array are 3 **undefined**

**I will try to explain it now:**

When you put the whole logic inside a function, in your case **getData** and pass the parameter **index**, you are applying the **lexical scope**.
It will remember the environment and the values(here **index**) from where it has been called in.
Which means the exact value of **index** will be preserved during the execution of this function. Say it was called with **index** 0 in first iteration. When the callback for **http.get** gets called for this iteration, the **index** will be 0. Same will happen for other iterations, it will correctly remember the value passed to it. 

For the wrong code, we were calling the logic directly from within the **For loop**. As we know the **For loop** will be executed sequentially and ends with a final state where **index** will become 3. 
Note that, till this point, the callbacks might not have been executed at all. Now when they start to get executed, they will seek the value of **index** and it will be 3. This is not right as you know for a 3 item array, the index were suppose to be 0, 1 and 2. Even at the end we will be printing out the 0, 1 and 2 indexed item of the **results** array.

So having an lexical scope by using a function in this scenario solves the problem.

Let me know if you have any question.

**Comment by ttchopper** [comment]
And Here is Mine.
   
    var http = require("http");
    var urls = process.argv.slice(2);
    var allData = [];
    var count = 0;

    urls.forEach(function(url, index) {
        http.get(url, function(response) {

            response.setEncoding("utf8");
            var data = "";

            response.on("data", function(chunk) {
                data += chunk;
            });
            response.on("end", function() {
                allData[index] = data;
                count++;
                if (count === urls.length) {
                    allData.forEach(function(data) {
                        console.log(data);
                    });
                }
            });
        });
    });

**Comment by radeng915** [comment]
I thought mine was pretty simple and practical. I'm pretty new, and I'd love some critique. All I did was split my http.get and my response callback into two seperate functions and basically just called the http.get whenever the first has finished. Set up an iterator so that it doesn't loop infinitely. It seems straightforward to me for any chunks of code that you need to run synchronously you should just split the function and call the first function again when the callback ends with some logical exit. Is this bad coding practice?

var http=require('http');
var urls=process.argv.slice(2,5);
var i=0;

function httpGet(url){
    if (i>2){return}
         else{   http.get(url,respond);}
}

function respond(response){
    var newData="";
    response.on("data",function(data){
    newData=newData+data;
    });
    response.on("end",function(){
        console.log(newData);
        i++;
        httpGet(urls[i]);
    });
}

httpGet(urls[i]);

**Comment by ishtiaque23** [comment]
@radeng915 


Nice piece of code. Does solve the problem. But it may not be the ideal solution we are looking for.

## Find reason by example:

Think of 3 http calls that require different time to execute:

1. http.get(url1) => (takes 50ms)
2. http.get(url2) => (takes 90ms)
3. http.get(url3) => (takes 30 ms)

**Case 1:**  If we run these synchronously, total execution time will be 170ms.
Why? Cuz they will wait for the earlier one to finish first. So, execution time will be sum of them.

**Case 2:** If we run these asynchronously, total execution time will be ~90ms.
Why? Cuz they will run in parallel and start almost at the same time. 
So by the time the 2nd call is done, it is expected the other asynchronous calls are finished already.

## The realization:
The 

> just called the http.get whenever the first has finished

 concept is considered to be **synchronous**. So your solution is a synchronous solution.

## Main focus of this exercise:
What this specific exercise **Juggling Async** want to teach us here is how we can get the most benefit out of the asynchronous calls in real life. But as some real world scenario might require, it asks for one extra thing. Do it while maintaining orders. 

1. An asynchronous solution will let us execute multiple call in parallel
2. It will take the time of the single longest running task.
3. But the response will come in an unordered fashion. You can not predict which comes when.

In this problem, you need to tackle that 3rd point while staying in the asynchronous realm.

Simplest Algorithm:
* Let them run asynchronously
* Store the response received in the callback somewhere(keep record of the order while doing so)
* Once all responses are in, print them(if you have maintained order in step 2, it will be perfect)

Let me know if this finding helps you. Would be glad to answer any other query.

**Comment by kalinchernev** [comment]
Thumbs up for the exercise, and in the spirit of the long thread of sharing :)
```
urls.forEach((url, index) => {
  (function(i){
    http.get(url, (res) => {
      res.setEncoding('utf8');
      var resContents = '';
      res.on('data', (chunk) => {
        resContents += chunk;
      });
      res.on('end', () => {
        result[i] = resContents;
        if (i == (urls.length - 1)) {
          result.forEach((ressult, j) => {
            console.log(ressult);
          });
        }
      });
      res.on('error', (error) => {
        throw new Error(error);
      });
    });
  }(index));
});
```

The discussions around the problem are pretty good, at least in the beginning where I had the patience to read :D

**Comment by wadetgong** [comment]
I got my code to work, but I'm not sure about one part. in the if(i === length-1) statement, I have var j iterate through the results array. I originally had var i in the for loop, but that led to an ECONNREFUSED error. I'm guessing it's because declaring var i = 0 in the for loop messes with httpIter(i+1,length), but I thought this instance of var i was supposed to be contained within the for loop?

`var http = require('http');
var bl = require('bl');
var results = [];

httpIter = function(i, length){
	http.get(process.argv[i], function(response){
		response.pipe(bl(function(err,data){
			results.push(data.toString());
				if(i === length-1){
					for(var j = 0;j<3; j++){
						console.log(results[j]);
					}
				}else{
					httpIter(i+1,length);
				}


		}));



	});
	

}
httpIter(2,process.argv.length);`

**Comment by Ardakilic** [comment]
This is my version, using recursive function:

```
var http = require('http');
var links = [2, 3, 4]; //process.argv orders. can easily be looped through process.argv.length though. Total laziness.
var buffer = [];
function render(index) {
    http.get(process.argv[links[index]], function (response){
        response.setEncoding('utf8');
        response.on('data', function(chunk){
            if(buffer[index] === undefined) {
                buffer[index] = '';
            }
            buffer[index] += chunk;
        });
        response.on('end', function () {
            var newIndex = index+1;
            if(links[newIndex] !== undefined) {
                render(newIndex);
            } else {
                return renderOutput();
            }
        });
        response.on('error', console.error);
    }).on('error', console.error);
}

function renderOutput() {
    buffer.forEach(function (elem) {
        console.log(elem);
    });
}

render(0);
```

I'm still new to this. so I hope this is correct and understandable 🙏

**Comment by JohnathanWhite** [comment]
Im sure this could be better but this is what I came up with.. thoughts?

const http = require('http');
const q = require('q');

const url1 = process.argv[2];
const url2 = process.argv[3];
const url3 = process.argv[4];

function retrieve(url){
    let deferred = q.defer();
    let dataString = '';
    let send = '';
    http.get(url, function(res){

        res.on('data', function(chunk){
            dataString += chunk
        });

        res.on('end', function(){
            deferred.resolve(dataString);
        });

        res.on('error', function(err){
            console.log(err);
            deferred.reject(err);
        });

    })
    return deferred.promise;
};

retrieve(url1).then(function(data){
    console.log(data)
    retrieve(url2).then(function(data){
        console.log(data)
        retrieve(url3).then(function(data){
            console.log(data)
        });
    });
});

**Comment by skoocda** [comment]
^ Looks good! 

You could generalise your import and final statement: try iterating through argv (with an offset of 2), to get all the URL values in a single object. Then you can loop through each to retrieve and log, i.e.

    for ( let url of urls ){
        retrieve( url ).then( function( data ){
           console.log(data);
        });
    }

**Comment by kfroemming** [comment]
Here is a shorter version that can be applied to any amount of url's for Http get. 

```
const http = require('http');
const bl = require('bl');
const sources = process.argv.slice(2);
let count = 0;
let content = [];

sources.forEach((source, index) => {
  http.get(sources[index], function(response) {
    response.pipe(bl(function (error, data) {
      if (error) {
        return console.error(error);
      }
      content[index] = data.toString();
      if (++count == sources.length) {
        content.forEach((el) => console.log(el));
      }
    }))
  })
})```

**Comment by ahmdshbz** [comment]
Hi,

I competed the exercise with the following program:

`var http = require('http')
var url1 = process.argv[2]
var url2 = process.argv[3]
var url3 = process.argv[4]
var data1 = ''
var data1_done = false
var data2 = ''
var data2_done = false
var data3 = ''
var data3_done = false

http.get(url1, gotHttp1).on('error', console.error)
http.get(url2, gotHttp2).on('error', console.error)
http.get(url3, gotHttp3).on('error', console.error)

function gotHttp1 (response) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    data1 += data
  })
  response.on('end', function () {
    data1_done = true
    tryToPrint()
  })
}

function gotHttp2 (response) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    data2 += data
  })
  response.on('end', function () {
    data2_done = true
    tryToPrint()
  })
}

function gotHttp3 (response) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    data3 += data
  })
  response.on('end', function () {
    data3_done = true
    tryToPrint()
  })
}

function tryToPrint () {
  if (data1_done && data2_done && data3_done) {
    console.log(data1)
    console.log(data2)
    console.log(data3)
  }
}`

I,m trying to improve the program with the fpllowing, which doesn't worl:
`var http = require('http')
var url = [process.argv[2], process.argv[3], process.argv[4]]
var dataBuf = ['', '', '']
var dataDone = [false, false, false]

getHttp()

function getHttp () {
  for (var i = 0; i < 3; ++i) {
    http.get(url[i], (response, i) => gotHttp).on('error', console.error)
  }
}

function gotHttp (response, i) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    console.log(data)
    dataBuf[i] += data
  })
  response.on('end', function () {
    dataDone[i] = true
    tryToPrint()
  })
}

function tryToPrint () {
  if (dataDone[0] && dataDone[1] && dataDone[2]) {
    for (var i = 0; i < 3; ++i) {
      console.log(dataBuf[i])
    }
  }
}
`

How can I send the index `i` as an extra parameter to `gotHttp` on line 10 of the second version?

**Comment by ahmdshbz** [comment]
Hi,

I competed the exercise with the following program:

```
var http = require('http')
var url1 = process.argv[2]
var url2 = process.argv[3]
var url3 = process.argv[4]
var data1 = ''
var data1_done = false
var data2 = ''
var data2_done = false
var data3 = ''
var data3_done = false

http.get(url1, gotHttp1).on('error', console.error)
http.get(url2, gotHttp2).on('error', console.error)
http.get(url3, gotHttp3).on('error', console.error)

function gotHttp1 (response) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    data1 += data
  })
  response.on('end', function () {
    data1_done = true
    tryToPrint()
  })
}

function gotHttp2 (response) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    data2 += data
  })
  response.on('end', function () {
    data2_done = true
    tryToPrint()
  })
}

function gotHttp3 (response) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    data3 += data
  })
  response.on('end', function () {
    data3_done = true
    tryToPrint()
  })
}

function tryToPrint () {
  if (data1_done && data2_done && data3_done) {
    console.log(data1)
    console.log(data2)
    console.log(data3)
  }
}
```

Then I unsuccessfully tried to make an improved second version:

```
var http = require('http')
var url = [process.argv[2], process.argv[3], process.argv[4]]
var dataBuf = ['', '', '']
var dataDone = [false, false, false]

getHttp()

function getHttp () {
  for (var i = 0; i < 3; ++i) {
    http.get(url[i], (response, i) => gotHttp).on('error', console.error)
  }
}

function gotHttp (response, i) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    console.log(data)
    dataBuf[i] += data
  })
  response.on('end', function () {
    dataDone[i] = true
    tryToPrint()
  })
}

function tryToPrint () {
  if (dataDone[0] && dataDone[1] && dataDone[2]) {
    for (var i = 0; i < 3; ++i) {
      console.log(dataBuf[i])
    }
  }
}
```

How do I pass the index `i` as an extra argument to `gotHttp`?

**Comment by ahmdshbz** [comment]
Hi,

I competed the exercise with the following program:

```
var http = require('http')
var url1 = process.argv[2]
var url2 = process.argv[3]
var url3 = process.argv[4]
var data1 = ''
var data1_done = false
var data2 = ''
var data2_done = false
var data3 = ''
var data3_done = false

http.get(url1, gotHttp1).on('error', console.error)
http.get(url2, gotHttp2).on('error', console.error)
http.get(url3, gotHttp3).on('error', console.error)

function gotHttp1 (response) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    data1 += data
  })
  response.on('end', function () {
    data1_done = true
    tryToPrint()
  })
}

function gotHttp2 (response) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    data2 += data
  })
  response.on('end', function () {
    data2_done = true
    tryToPrint()
  })
}

function gotHttp3 (response) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    data3 += data
  })
  response.on('end', function () {
    data3_done = true
    tryToPrint()
  })
}

function tryToPrint () {
  if (data1_done && data2_done && data3_done) {
    console.log(data1)
    console.log(data2)
    console.log(data3)
  }
}
```

Then I unsuccessfully tried to make an improved second version:

```
var http = require('http')
var url = [process.argv[2], process.argv[3], process.argv[4]]
var dataBuf = ['', '', '']
var dataDone = [false, false, false]

getHttp()

function getHttp () {
  for (var i = 0; i < 3; ++i) {
    http.get(url[i], (response, i) => gotHttp).on('error', console.error)
  }
}

function gotHttp (response, i) {
  response.setEncoding('utf8')
  response.on('error', console.error)
  response.on('data', function (data) {
    console.log(data)
    dataBuf[i] += data
  })
  response.on('end', function () {
    dataDone[i] = true
    tryToPrint()
  })
}

function tryToPrint () {
  if (dataDone[0] && dataDone[1] && dataDone[2]) {
    for (var i = 0; i < 3; ++i) {
      console.log(dataBuf[i])
    }
  }
}
```

How do I pass the index `i` as an extra argument to `gotHttp`? (on line 10 of the second version)

**Comment by kfroemming** [comment]
@ahmdshbz 
``` 
10     http.get(url[i], (response) => gotHttp(response, i)).on('error', console.error)
```
The line above will work, you just want to pass response first, then call it in the callback and pass both parameters. I don't think your code is going to pass though. Look at my example posted above for a good idea of what you can do ti improve it.

**Comment by ahmdshbz** [comment]
@kfroemming 
Thanks for your reply

I tried to modify your example according to the guidelines given at [www.callbackhell.com](url)
I named the anonymous functions and moved them to the top level of the program, but the modification didn't work:
```
const http = require('http');
const bl = require('bl');
const sources = process.argv.slice(2);
let count = 0;
let content = [];

sources.forEach(processUrl);

function processUrl (source, index) {
  http.get(sources[index], function (response) {
    response.pipe(bl(processBuffer));
  });
}

function processBuffer (error, data) {
  if (error) {
    return console.error(error);
  }
  content[index] = data.toString();
  if (++count == sources.length) {
    content.forEach((el) => console.log(el));
  }
}
```

**Comment by noffle** [comment]
@ahmdshbz What is the value of `index` inside of `processBuffer`? How is `index` getting into that function? What value should it have?

**Comment by ahmdshbz** [comment]
@noffle 
Thanks for your comment

I think the value of `index ` inside of `processBuffer` should match the index of the URLs in `sources`
So `index` should be passed to `processBuffer`...

```
const http = require('http');
const bl = require('bl');
const sources = process.argv.slice(2);
let count = 0;
let content = [];

sources.forEach(processUrl);

function processUrl (source, index) {
  http.get(sources[index], function (response) {
    response.pipe(bl((err, data) => processBuffer(err, data, index)));
  });
}

function processBuffer (error, data, index) {
  if (error) {
    return console.error(error);
  }
  content[index] = data.toString();
  if (++count === sources.length) {
    content.forEach((el) => console.log(el));
  }
}
```

It works! Thanks a lot!!

**Comment by BryceLarkin** [comment]
@grantcarthew Working on this right now and your code was very helpful in helping me understand the problem. I think the count is off by one though. This seems to work:
```javascript
var http = require('http');
var urls = process.argv.slice(2, process.argv.length);
var urlsCount = urls.length


var results = [];
var waiting = urlsCount;

urls.map(function(url, index) {
  http.get(url, function(resp) {
    var alldata = "";
    resp.setEncoding('utf8');
    resp.on('data', function(data){alldata += data});
    resp.on('end', function() {
      results[index] = alldata;
      waiting = waiting - 1;
      if (!waiting) {
        results.map(function(result){console.log(result)});
      }
    });
  });
});
```

**Comment by grantcarthew** [comment]
Hi @BryceLarkin. Wow blast from the past. That was a long time ago. I would be writing the code very differently now. I would be refactoring all the functions out to either named functions or anonymous functions. This would make the code a lot easier to read. Your code looks good though.

**Comment by BalaGanesh555** [comment]
![Screenshot (9)](https://user-images.githubusercontent.com/59443594/71795386-f2b7ff00-306b-11ea-826a-23d2a532fdfd.png)
 would you please sole this
